Mag l1a compression

maxinelasp opened the pull request:
# Change Summary
This is a pretty involved PR. I would like to thank in advance anyone who reviews. I don't think breaking this code up further would make sense, or I would have done so.

I'd recommend starting by reading the docstring for "process_compressed_vectors" in mag_l1a_data.py, as it gives a good overview on the overall algorithm.

Note: "process_uncompressed_vectors" was moved into a new method but otherwise unchanged. 

## Overview
<!--Add a list or paragraph giving an overview of your changes-->
- Added MAG decompression option
- Added tests

## Updated Files
<!--List each updated file and add bullets to describe the purpose/function of the updates.
This should be high level, but enough detail for the reviewer to understand what purpose(s) the
updated code in the file is serving-->
- mag_l1a_data.py
   - Added 7 new methods to decode compressed data

- test_mag_l1a
   - Added tests
- mag_l0_data
  - Minor fix for mypy

## Testing
<!--List any relevant testing information like number of unit tests added, any new tests being skipped and the reason why, etc.-->
Testing is pretty extensive for this change, since it's so complex. I generated test data primarily using a piece of code that the MAG team provided for me.

Closes issue #727

maxinelasp committed:
Setting up compressed vector processing

maxinelasp committed:
First pass at compressed processing, and tests

maxinelasp committed:
First pass at compression

maxinelasp committed:
some tidying up

maxinelasp committed:
Updating to include range data section processing/tests

maxinelasp committed:
Attempting to add uncompression steps

maxinelasp committed:
got it working

maxinelasp committed:
Finished compression algorithm

maxinelasp committed:
Merge branch 'upstream-dev' into mag_l1a_compression

maxinelasp committed:
Updating ruff and mypy errors

maxinelasp committed:
removing unneeded changes

maxinelasp commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

maxinelasp committed:
Updating docs

maxinelasp committed:
Merge branch 'mag_l1a_compression' of github.com:maxinelasp/imap_processing into mag_l1a_compression

maxinelasp committed:
Fixing error with update from MAG team

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +477,463 @@ def to_signed16(n: int) -> int:
+        # The first 8 bits are a header - 6 bits to indicate the compression width,
+        # 1 bit to indicate if there is a range data section, and 1 bit spare.
+        compression_width = int("".join([str(i) for i in bit_array[:6]]), 2)
+        has_range_data_section = int(str(bit_array[6]), 2)
```
```suggestion
        has_range_data_section = (bit_array[6] == 1)
```
be nicer if this was typed as a bool?

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +477,463 @@ def to_signed16(n: int) -> int:
+        # the range
+        uncompressed_vector_size = compression_width * 3 + 2
+        # plus 8 to get past the compression width and range data section
+        first_vector_width = uncompressed_vector_size + 8
```
I dont understand the +8 here - you are aready skipping 8 bits in bit array on line 546?

maxinelasp replied:
This is the index of the end of the first vector, so it's uncompressed_vector size long, but we need to skip past the first 8 bits - so the end index is uncompressed_size plus 8. I can move the `+8` into the slice in line 546, if that makes more sense?

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +477,463 @@ def to_signed16(n: int) -> int:
+
+        # The full vector includes 3 values of compression_width bits, plus 2 bits for
+        # the range
+        uncompressed_vector_size = compression_width * 3 + 2
```
Would be tempted to 3 as a constant NUMBER_OF_AXIS and 2 as BIT_WIDTH_OF_VECTOR_RANGE for reader sanity

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +477,463 @@ def to_signed16(n: int) -> int:
+
+        end_vector = (
+            len(bit_array)
+            - (primary_count + secondary_count) * 2 * has_range_data_section
```
range_data_section has one less range than total as first range is in the header

maxinelasp replied:
Good catch! Fixed this

maxinelasp committed:
Fixing compressed data errors with range

tech3371 commented on imap_processing/mag/constants.py:
```diff
@@ -55,3 +55,49 @@ class PrimarySensor(Enum):
+    165580141,
+]
+
+MAX_FINE_TIME = 65535  # maximum 16 bit unsigned int
```
?
```suggestion
MAX_FINE_TIME = np.iinfo(np.uint16).max 
```

maxinelasp replied:
This change actually randomly fixed a mypy error 🤷  so thanks for pointing it out!

tech3371 commented on imap_processing/mag/l1a/mag_l1a.py:
```diff
@@ -42,6 +42,10 @@ def mag_l1a(packet_filepath: Path, data_version: str) -> list[xr.Dataset]:
     norm_data = packets["norm"]
     burst_data = packets["burst"]
 
+    print(np.unpackbits(norm_data[0].VECTORS))
+    print(len(np.unpackbits(norm_data[0].VECTORS)))
+    print(len(norm_data[0].VECTORS))
```
```suggestion
```

maxinelasp replied:
Good find, thank you!

tech3371 commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -287,10 +296,50 @@ def calculate_vector_time(
+        -------
+        (primary, secondary): (numpy.ndarray, numpy.ndarray)
+            Two arrays, each containing tuples of (x, y, z, sample_range) for each
+            vector sample.
```
very helpful comments!

tech3371 commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +477,467 @@ def to_signed16(n: int) -> int:
+        Given raw compressed packet data, process into Vectors.
+
+        To do this, we need to decode the compressed data. The compressed data starts
+        with an 8 bit header that defines the width of the uncompressed vectors and
```
I think we decided to use decompressed instead of uncompressed. Can you change it in other places as well?
```suggestion
        with an 8 bit header that defines the width of the decompressed vectors and
```

maxinelasp replied:
I would use "decompressed" to mean the data after I had finished with the algorithm to turn it from compressed -> no longer compressed. In this case, the data was never compressed to start. I can change it if that's what we have decided "decompressed" means, but I want to call that out. 

Also, MAG uses "uncompressed" which is where that word came from. Ultimately, the decompressed/uncompressed word would only be used in this code and not in any of the outputs - so while I normally would be all the way for using the same word, in this case, I'd prefer to match the MAG team and not the other instruments.... but I am definitely open to being convinced. Where else are we using "decompressed?"

tech3371 replied:
I don't feel strongly about this since it's just in the doc string. I did look at your function with name `uncompressed` and it mades sense to keep it `uncompressed`. May be it's better to keep the comment to match that. Ok, ignore my comment since the data is not compressed.

maxinelasp replied:
Cool, thanks for bringing it up! It's good to get the check to make sure my words are good 😄

tech3371 commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +477,467 @@ def to_signed16(n: int) -> int:
+        uncompressed primary vector with range. Then, we proceed through the compressed
+        data, where each value is fibonacci and zig-zag encoded. This means each value
+        ends in 2 sequential ones (11). We split the data along these numbers until
+        we reach primary_count.
```
Do you mean this?
```suggestion
        we reach last primary_count.
```

maxinelasp replied:
This is an awkward sentence, and I will rephrase it, thanks! It should be "until we reach primary_count vectors"

tech3371 commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +477,467 @@ def to_signed16(n: int) -> int:
+        """
+        bit_array = np.unpackbits(vector_data)
+        # The first 8 bits are a header - 6 bits to indicate the compression width,
+        # 1 bit to indicate if there is a range data section, and 1 bit spare.
```
These doc strings and comments are helping a lot to follow with your code.

maxinelasp replied:
Good! I'm glad. I would have been lost without them 😅

tech3371 commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +477,467 @@ def to_signed16(n: int) -> int:
         )
+
+    @staticmethod
+    def process_compressed_vectors(  # noqa: PLR0912, PLR0915
```
I would really feel good if MAG reviewed this function. I am only understanding small portion and very high level. I apologize that I am not much of help here.

maxinelasp replied:
Thank you for looking, you have been helpful! I have asked them for a review, particularly of the tests.

maxinelasp committed:
PR comments

maxinelasp committed:
fixing mypy version and error

maxinelasp committed:
Mypy change

laspsandoval commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +477,467 @@ def to_signed16(n: int) -> int:
         )
+
+    @staticmethod
+    def process_compressed_vectors(  # noqa: PLR0912, PLR0915
```
Is there a way to break this up at all?

maxinelasp replied:
It's already fairly broken up, personally I don't think there are other obvious things to pull out. But, if you'd prefer, I can break it up further. I know it's a bit of a monster method.

alastairtree replied:
Yeah it is hard to get it more split up, but still probably a worth endeavour if you can as it is very long.

laspsandoval commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +477,467 @@ def to_signed16(n: int) -> int:
+        -------
+        value: int
+            Signed integer value, with fibonacci and zig-zag encoding removed.
+        """
```
Could there be a case in which code is too short?
if len(code) < 2:
        raise ValueError(f"Input code too short: {code}")

maxinelasp replied:
I think it is ok for the length to be equal to 2 (this is how you represent 0.) If it's shorter than two, then it will be caught with the existing check... although without the nice error message. I'll add that in, good suggestion

laspsandoval commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +477,467 @@ def to_signed16(n: int) -> int:
+            Two arrays, each containing tuples of (x, y, z, sample_range) for each
+            vector sample.
+        """
+        bit_array = np.unpackbits(vector_data)
```
Is there a reason to pack the bits and then unpack them in the same program?

maxinelasp replied:
The packing in the tests is to get the test data to line up with the real output data. It's basically convenient to have the test data explicitly in bits, and then pack that into bytes to match what the expected packet data looks like, because the bytes are very obscure. In the actual steps, this is the only unpack step, and then later they are re-packed in the uncompressed case, but it's still more convenient to use the output of `np.unpack`. 

Long story short, in the actual data, the data is already packed, so there isn't a packing then unpacking step like there is in the test data.

laspsandoval commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +477,467 @@ def to_signed16(n: int) -> int:
+        has_range : int
+            1 if the vector data includes range data, 0 if not. The first vector always
+            has range data, it is only if the compression fails and we revert to
+            uncompressed data partway through that we skip the range.
```
So would it either be one of three scenarios:
1. compressed - range data
2. compressed + uncompressed - range data
3. uncompressed - no range data

What do you mean by "partway through"? This is more for my own understanding than anything.

maxinelasp replied:
You're correct. I'll reword this sentence because I understand your confusion. The `has_range` section is totally separate from the compressed/uncompressed bits. What this sentence is trying to say, is that the uncompressed data does not include a range value. But that should probably be said somewhere else. 

It is also possible to have compressed + uncompressed with no range data. If the range stays the same for the entire packet, then we do not have a range data section. Otherwise, we retrieve all the range data for all the vectors from the range data, regardless if the vectors are compressed or uncompressed. 

I'll rewrite this sentence because it is confusingly tying things together when they should be separate, thanks for calling that out!

laspsandoval reviewed (commented):
Interesting concept. Only a few suggestions.

maxinelasp committed:
PR updates

maxinelasp committed:
Fixing issues now that we have real data - temporary update

vmartinez-cu commented:
Nice job with your doc strings and comments. The details are helpful

sdhoyt commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+        if there is a range data section. Then, the vector data follows, then the range
+        data section if it exists.
+
+        To decode, we start by decoding the first compression_width bits. This is an
```
>To decode, we start by decoding the first compression_width bits. This is an
         uncompressed primary vector with range.

I'm having trouble understanding this part. Looking at the code, the compression_width is uncompressed 6 bits of the header. What does it mean that this is the "uncompressed primary vector with range"? Is this giving the bit length of the compressed data in the primary vector?

maxinelasp replied:
Yeah, sorry, this sentence is kind of weird. The compression width is the first 6 bits of the header, which corresponds to a bit count that indicates the width of uncompressed vectors. 

So, if that compression width was `01 0000`, then we have uncompressed vectors values of a width of 16 bits. The full width of the first vector is 16 * 3 + 2 bits, because there are three values per vector and 2 bits for the range value.

alastairtree replied:
Say header compression_width=16, then you know the first vector is length (x=16bits,y=16bit,z=16bits,rng=2bits)=50bits
After that the rest is compressed with varying bit width, but will decompress into 16bit x/y/z values.

sdhoyt commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+
+        To do this, we need to decode the compressed data. The compressed data starts
+        with an 8 bit header that defines the width of the uncompressed vectors and
+        if there is a range data section. Then, the vector data follows, then the range
```
It looks like range data is a kind of MAG data that may or may not be present,  not something used specifically for the decompression scheme, correct?

maxinelasp replied:
Yes, the range data is the last piece of information about the vector. In the compressed case, we assume that the range is the same for all vectors unless there is a range data section, in which case we read the range data section at the end to update the ranges.

alastairtree replied:
Each sensors are in one of 4 ranges 0,1,2,3 at any given moment, based on the size of the field. A tiny field would be range 3, and then when a big flare passes imap it might switch into range 2 for a while, and then go back to range 3. On earth is range 0. 
99% of the time it does not change range so you can assume every vector is in the same range as the first vector. When it switches range you need to know that say the first 5 vectors were in range 3, and vectors 6-128 are in range 2 so we include the range array only when you need that data as an extra way to save bandwidth.

alastairtree commented on imap_processing/tests/mag/test_mag_l1a.py:
```diff
@@ -47,7 +305,364 @@ def test_compare_validation_data():
+    headers = "01000010"
+    # TODO: if there is a number of vectors that is not a multiple of 8,
+    #  is there also padding at the end of the range section?
+    input_data = np.array(
```
Yes the range section ends on the byte boundary with padding if needed but as it is the end of the packet it can be ignored. Guess this TODO can be removed/resolved.

alastairtree commented on imap_processing/tests/mag/test_mag_l1a.py:
```diff
@@ -47,7 +305,364 @@ def test_compare_validation_data():
+    assert secondary.shape[0] == 16
+
+    assert primary.shape[1] == 4
+    assert secondary.shape[1] == 4
```
The 16 and 4 here are a bit magical. I think you are checking 16 vectors per packet, and 4 fields (x,y,z,r). Maybe add a comment? Also I think this is kinda covered already by the `np.array_equal` assertion above?

alastairtree commented on imap_processing/tests/mag/test_mag_l1a.py:
```diff
@@ -47,7 +305,364 @@ def test_compare_validation_data():
+        primary_expected[i][3] = expected_range_primary[i]
+        secondary_expected[i][3] = expected_range_secondary[i]
+
+    (primary_with_range, secondary_with_range) = MagL1a.process_compressed_vectors(
```
I think this test should be broken out into 2 separate tests, one for with range data and one without. The one without should check the generated range data all matches the first vectors range.

alastairtree commented on imap_processing/tests/mag/test_mag_l1a.py:
```diff
@@ -47,7 +305,364 @@ def test_compare_validation_data():
+    # It excludes the first vector, making the length (primary_count - 1) * 2
+    range_primary = "000000000000000101101011111111"
+
+    expected_range_primary = [3, 0, 0, 0, 0, 0, 0, 0, 1, 1, 2, 2, 3, 3, 3, 3]
```
```suggestion
    first_primary_vector_range = 3  # This is not included in the range data section
    expected_range_primary = [first_primary_vector_range, 0, 0, 0, 0, 0, 0, 0, 1, 1, 2, 2, 3, 3, 3, 3]
```
Just for readability

alastairtree commented on imap_processing/tests/mag/test_mag_l1a.py:
```diff
@@ -47,7 +305,364 @@ def test_compare_validation_data():
+        "000000100000110100001000001101110001000001101111"
+    )
+
+    # TODO fix test
```
Still To do?

alastairtree commented on imap_processing/tests/mag/test_mag_l1a.py:
```diff
@@ -47,7 +305,364 @@ def test_compare_validation_data():
+    assert np.array_equal(primary[16:], uncompressed_expected_vectors)
+
+    assert np.array_equal(secondary[:15], expected_vectors[1][:-1])
+    assert np.array_equal(secondary[16:], uncompressed_expected_vectors)
```
Both primary and secondary have the same expected vectors for 16th onwards?

maxinelasp replied:
Yes, I just used the same bits for primary and secondary

alastairtree commented on imap_processing/tests/mag/test_mag_l1a.py:
```diff
@@ -47,7 +305,364 @@ def test_compare_validation_data():
+    (primary, secondary) = MagL1a.process_compressed_vectors(input_data, 6, 16)
+    assert len(primary) == 6
+    assert np.array_equal(primary[0], expected_vectors[0][0])
+    assert np.array_equal(primary[2:], uncompressed_expected_vectors)
```
This seems like a second test and I can't work out what it is testing? What does "Test if first primary vector is too long" mean?

maxinelasp replied:
This is checking some weird edge cases in my code. The first part of the test is checking to see if a vector partway through is > 60 bits, this part checks if the first vector is >60 bits.

alastairtree commented on imap_processing/tests/mag/test_mag_l1a.py:
```diff
@@ -47,7 +305,364 @@ def test_compare_validation_data():
+
+def test_different_compression_width():
+    # Compression headers - indicating a 12 bit width and no range section
+    headers = "00110000"
```
Would be tempted to have a method to build this to make it readable and reduce duplication:
```python
headers=generate_mag_science_header(compression=False, bit_width=12)
```

alastairtree commented on imap_processing/tests/mag/test_mag_l1a.py:
```diff
@@ -47,7 +305,364 @@ def test_compare_validation_data():
+
+    expected_vectors = np.array(
+        [[1, 2, 3, 4], [2, 3, 4, 4], [5, 3, 1, 4], [4, -7, 2, 4]]
+    )
```
```python
   range=3 
   start_vector = np.array([1, 2, 3, range], dtype=np.uint)

    diff_vectors = [
        1, 1, 1, 
        3, 0, -3, 
       -1, -10, 1]

    expected_vectors = np.array(
        [[1, 2, 3, range], [2, 3, 4, range], [5, 3, 1, range], [4, -7, 2, range]]
    )
```
Something like this makes it easier to read/review?

alastairtree commented on imap_processing/tests/mag/test_mag_l1a.py:
```diff
@@ -47,7 +305,364 @@ def test_compare_validation_data():
+        [[1, 2, 3, 4], [2, 3, 4, 4], [5, 3, 1, 4], [4, -7, 2, 4]]
+    )
+
+    test_vectors = MagL1a.accumulate_vectors(start_vector, diff_vectors, 4)
```
Would be tempted to rename this something like `convert_vectors_from_diffs_to_eng_units()` but naming is hard and i have a preference for long_but_very_readable_method_names :-)

maxinelasp replied:
I don't love this variable name, but your suggestion is confusing to me, because that is not how I think about this method. But, I will rename it

alastairtree commented on imap_processing/tests/mag/test_mag_l1a.py:
```diff
@@ -47,7 +305,364 @@ def test_compare_validation_data():
+    assert np.array_equal(test_vectors, expected_vectors)
+
+
+def test_unpack_one_vector(uncompressed_vector_bytearray, expected_vectors):
```
Might be better as a parameterised test rather than multiple tests in one. Our vectors will always be between 12 and 20 bits when compressed, and always 16bit when uncompressed

alastairtree commented on imap_processing/tests/mag/test_mag_l1a.py:
```diff
@@ -61,7 +676,7 @@ def test_process_vector_data():
     # 36 bytes
     (primary_vectors, secondary_vectors) = MagL1a.process_vector_data(
-        input_data, total_primary_vectors, total_secondary_vectors
+        input_data, total_primary_vectors, total_secondary_vectors, 0
```
Maybe use a constant or enum here
```suggestion
        input_data, total_primary_vectors, total_secondary_vectors, UNCOMPRESSED
```

alastairtree commented on imap_processing/tests/mag/test_mag_l1a.py:
These tests look good thanks, have added a few inline style things. 
Missing cases I can see it might be worth covering with tests:
- When there is no range section and so all vectors in the same range based on the first vector
- having different rates in primary and secondary, like in Burst(64,8) mode say
- different cases for the padding between sections, like when the end of the primaries falls on the byte boundary and so there is no padding

maxinelasp replied:
Your first case is covered under the first part of `test_compressed_vector_data`, but your other tests are good and I will definitely add them, thanks!

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+                secondary_vectors,
+            )
+
+        return primary_vectors, secondary_vectors
```
I would be tempted to put in some defensive assertions here? E.g. that you have the desired number of pri/sec vectors, and that each has a valid range 0-3.

maxinelasp replied:
Sure, I can add those checks

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+                vector_count > primary_count
+                and seq_val - secondary_boundaries[-1] > 0
+            ):
+                secondary_boundaries.append(seq_val + 1)
```
The logic for primary and secondary processing of boundaries is duplicated and a good candidate for refactoring out to a method because the primary and secondary sections are processed/packed the same way just with different start/end locations in the bit stream

maxinelasp replied:
Agreed, and I've had to update two places a few times now, so I'll pull this out into another method.

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+                np.roll(vector_bits, 1),
+            )[1:]
+            == 2
+        )[0]
```
This is a much fancier method to split the bits into individual vectors all in one go, nice! I just iterated the array and chopped after sequential 1s!

maxinelasp replied:
I dug deep into my days of hackerrank to get this weird algorithm 😅 in the end I'm not totally convinced it's better than your design.... but it is quite fast, which is good considering we will probably need to process this code a couple of times for the perfect 24 hour files.

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+                        (len(primary_boundaries) > 4)
+                        and (primary_boundaries[-1] - primary_boundaries[-4] > 60)
+                        or (vector_count == 2 and primary_boundaries[-1] > 60)
+                    ):
```
The 60 here should probably be refactored out to a constant EXIT_COMPRESSION_MAX_BITS_THRESHOLD or something.

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+                    if (
+                        (len(primary_boundaries) > 4)
+                        and (primary_boundaries[-1] - primary_boundaries[-4] > 60)
+                        or (vector_count == 2 and primary_boundaries[-1] > 60)
```
```
                       (len(primary_boundaries) > 4)
                        and (primary_boundaries[-1] - primary_boundaries[-4] > 60)
                        or (vector_count == 2 and primary_boundaries[-1] > 60)
```
I have read this a few times and cannot quite work out why len > 4 or count == 2 for this so i think it needs a refactor or explanation

maxinelasp replied:
This line caused me much trouble. The beginning and ends of the sequence are kind of weird, since each index indicates the end of the value. This means we need to compare the most recent value to 3 indices ago - except for the first vector, which does not have  3 prior indices, only 2. So in that case we just look at the index of the last value to see if it's > 60. 

I will update the comment to explain this a bit more, thanks for calling it out.

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+            # If we have 3 ones in a row, we should skip that index
+            if vector_count < primary_count and (
+                seq_val - primary_boundaries[-1] > 0
+            ):
```
huh?

maxinelasp replied:
with the np.roll method, the indices will be duplicated for 3 ones in a row. This is allowed, but we need to skip the indices for the second pair of ones. 

eg: [0, 1, 1, 1, 0, 0, 1, 1] is valid and should be split into [0, 1, 1] and [1, 0, 0, 1, 1], along the indices 1 and 6. 

However, there are also 2 sequential ones at index 3. So, the output of np.roll is [0, 1, 1, 1, 0, 0, 1, 1] + [1, 1, 1, 0, 0, 1, 1, 0] = [1, 2, 2, 1, 0, 1, 2, 1] with indices [1, 2, 6]. So we need to drop that 2.

I'll update the comment to be more clear. This is the main reason why I think this code is... not ideal.

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+                primary_boundaries.append(seq_val + 1)
+
+                # 3 boundaries equal one vector
+                if len(primary_boundaries) % 3 == 0:
```
```suggestion
                if len(primary_boundaries) % NO_OF_AXIS == 0:
```

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+        secondary_boundaries = []
+        vector_count = 1
+        end_primary_vector = 0
+        for seq_val in sequential_ones:
```
It might be easier to read/follow this if you iterated in jumps of three here, and then process the x/y/z values as a whole vector?

maxinelasp replied:
I'll give it a shot and see if that helps make things more readable!

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+                )
+                primary_vectors[vector_index] = decoded_vector
+                primary_vectors[vector_index][3] = first_vector[3]
+                vector_index += 1
```
Sorry I cant really follow this and don't understand why any vectors would ever be "missing"?

maxinelasp replied:
Yeah, bad phrasing. They aren't missing from the data, but rather missing from the compressed vectors. If we finish with the compressed vectors, and we don't have the number that we expect, then this code assumes that they are missing because they are in uncompressed form. So, it's more like they're missing from what has been processed so far. I will try to rename this and expand my comments.

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+                    + (primary_count + secondary_count - 2) * 2
+                ],
+                secondary_vectors,
+            )
```
Not really clear from this that if `has_range_data_section=0` then all vectors have the same range as their sensors first vector. Refactor the processing of the range section out to a method and make that logic explicit?

maxinelasp replied:
So, do you mean that I should always call `process_range_data_section` and pass `has_range_data_section` through, or should I just be more explicit within this if statement?

alastairtree replied:
I think maybe calling a method called `unpack_vector_ranges` that either copies it from the first vector, or reads the end section would be more explicit/easier to understand. ATM there are 2 really separate bits of code, (line 770 `vectors[i + compressed_count][3] = vectors[0][3]`) and here that generate the range for each vector, depending on the value of `has_range_data_section` and it is non-obvious that they are connected

alastairtree replied:
Also seems like the logic is on line 865 as well `vectors[vector_index][3] = vectors[0][3]`

maxinelasp replied:
Ok yeah, that makes sense. I will rework that, I think it will be more clear. Thanks!

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+        Parameters
+        ----------
+        range_data : numpy.ndarray
+            Array of range values, where each value is one bit.
```
```suggestion
            Array of range data section in binary. Each range is 2 bits long.
```

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+        range_str = "".join([str(i) for i in range_data])
+        for i in range(len(vectors) - 1):
+            range_int = int(range_str[i * 2 : i * 2 + 2], 2)
+            updated_vectors[i + 1][3] = range_int
```
It bugged me how this went from a array of bits to a string to ints. Best i could do to simplify was like this into array of 2bit range values. Will be faster but not sure if it is any better?

```python
arr = np.array([0,0,0,1,1,0,1,1]) #[0,1,2,3] as 2 bit numbers in binary array
result = [(x<<1)+y for x,y in np.array_split(arr,len(arr)/2)]
```

maxinelasp replied:
This was a slow bit in my profiling. But that change is less intuitive.... I'll contemplate it, thanks for the suggestion!

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+            logger.error(f"Error! Computed compressed vector differences are not "
+                         f"divisible by 3 - meaning some data is missing. "
+                         "Expected length: %s, actual length: "
+                         "%s", vector_count * 3, len(vector_differences))
```
If this happened we would probably say abandon the packet as something very bad has happened

maxinelasp replied:
I emailed you about this - this was originally an error, but I changed it to a log because the test packets you sent me had several packets where this error cropped up. 

In the future, would you want this to generate a CDF file that is empty, or skip generating the CDF file altogether? Skip that packet only, but include the other packets if they're ok? Pass in the data we have but mark a quality flag? Something else....?

Thanks for your review!

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+                         "%s", vector_count * 3, len(vector_differences))
+        index = 0
+        vector_index = 1
+        for diff in vector_differences:
```
Another readability thing but how about in chunks of 3 something like this?
```suggestion
        for vector in [(x,y,z) for x,y,z in np.array(vector_differences).reshape(-1,3)] :
```

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+                f"{width * 3} or {width * 3 + 2} if has_range."
+            )
+            return np.zeros(4, dtype=np.int32)
+            # TODO: should this raise an error? Update flags?
```
Should fail the whole packet if this errors

alastairtree commented on imap_processing/mag/l1a/mag_l1a_data.py:
```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+
+        range_string = "".join([str(i) for i in vector_data[-2:]])
+
+        rng = int(range_string, 2) if has_range else 0
```
This range logic is duplicated above and could be a method

alastairtree reviewed (commented):
Epic effort maxine, thank you!

I have left you lots of comments (sorry!). Lots are just style and readability stuff - i know how hard it is to make this compression logic comprehendable so I feel your pain.

Think my only other query is on error handling. I think it should be a bit more defensive and if you get any bad unexpected values it means something awful has happened and the whole packet should be considered a stream of NaNs and an error raised somewhere so we can investigate..

Thank you!

maxinelasp committed:
Fixing tests

maxinelasp committed:
Adding new tests

maxinelasp committed:
Pulling out duplicate code

maxinelasp committed:
PR updates

maxinelasp committed:
Merge branch 'upstream-dev' into mag_l1a_compression

maxinelasp committed:
correcting filler byte in tests

maxinelasp committed:
Correcting pre-commit

maxinelasp commented:
@alastairtree @tech3371 @sdhoyt @laspsandoval Thank you so much for your reviews! I have addressed all the comments on this PR, added some additional tests, and had the output validated by the MAG team. Could I get a re-review and maybe an approval from you?

sdhoyt reviewed (approved):
It looks good to me. Nice work!

alastairtree commented:
There are a couple of comments of mine that you "thumbs up'd" / replied to but didn't actually change, which is fine but might be worth checking you have pushed all your changes if not. Otherwise looks good!

maxinelasp commented:
@alastairtree Thanks for that final review! I will double check that I addressed everything, there were a few comments that I preferred to leave as-is, but I definitely want to make sure I decided on everything. If there are any comments that you definitely want to see changed, please let me know.

maxinelasp commented:
@alastairtree Thanks for pointing out those comments, I guess I resolved a bunch of them in some kind of fugue state on Friday afternoon and didn't actually fix the things you were pointing out! I went back in and I'm addressing your comments, including a bunch on the tests that I missed. Sorry about that, I got a little over eager to finish this up 😅

maxinelasp committed:
Addressing comments

maxinelasp committed:
Merge branch 'upstream-dev' into mag_l1a_compression

maxinelasp committed:
fixing poetry file

maxinelasp committed:
fixing tests + pre commit

maxinelasp committed:
Fix tests plus docs
